src/hotspot/os/bsd/os_perf_bsd.cpp

```diff
@@ -52,17 +52,150 @@
 
 static const double NANOS_PER_SEC = 1000000000.0;
 
+#if !defined(__APPLE__)
+struct CPUTicks {
+  uint64_t usedTicks;
+  uint64_t totalTicks;
+};
+
+struct JVMTicks {
+  uint64_t userTicks;
+  uint64_t systemTicks;
+  CPUTicks cpuTicks;
+};
+
+struct PerfCounters {
+  int nProcs;
+  int stathz;		/* statistics clock frequency */
+  JVMTicks jvmTicks;
+  CPUTicks* cpus;
+};
+
+static int get_stathz(int *stathz) {
+  struct clockinfo ci;
+  size_t length = sizeof(ci);
+  int mib[] = { CTL_KERN, KERN_CLOCKRATE };
+  const u_int miblen = sizeof(mib) / sizeof(mib[0]);
+
+  if (sysctl(mib, miblen, &ci, &length, NULL, 0) == -1) {
+    return OS_ERR;
+  }
+
+  *stathz = ci.stathz;
+
+  return OS_OK;
+}
+
+static int get_cpu_ticks(CPUTicks *ticks, int which_logical_cpu, int nProcs) {
+
+#if defined(__NetBSD__)
+uint64_t cpu_load_info[CPUSTATES];
+#else
+long cpu_load_info[CPUSTATES];
+#endif
+size_t length = sizeof(cpu_load_info);
+
+  if (which_logical_cpu == -1) {
+#if defined(__OpenBSD__)
+    int mib[] = { CTL_KERN, KERN_CPTIME };
+    const u_int miblen = sizeof(mib) / sizeof(mib[0]);
+
+    if (sysctl(mib, miblen, &cpu_load_info, &length, NULL, 0) == -1) {
+      return OS_ERR;
+    }
+    /* OpenBSD returns the sum/nProcs. Unify with other stat units */
+    for (size_t i=0; i < CPUSTATES; i++) {
+       cpu_load_info[i] *= nProcs;
+    }
+#else
+    if (sysctlbyname("kern.cp_time", &cpu_load_info, &length, NULL, 0) == -1) {
+      return OS_ERR;
+    }
+#endif
+  } else {
+#if defined(__OpenBSD__)
+    int mib[] = { CTL_KERN, KERN_CPTIME2, which_logical_cpu };
+    const u_int miblen = sizeof(mib) / sizeof(mib[0]);
+
+    if (sysctl(mib, miblen, &cpu_load_info, &length, NULL, 0) == -1) {
+      return OS_ERR;
+    }
+#elif defined(__FreeBSD__)
+    size_t alllength = length * nProcs;
+    long *allcpus = NEW_C_HEAP_ARRAY(long, CPUSTATES * nProcs, mtInternal);
+
+    if (sysctlbyname("kern.cp_times", allcpus, &alllength, NULL, 0) == -1) {
+      FREE_C_HEAP_ARRAY(long, allcpus);
+      return OS_ERR;
+    }
+
+    memcpy(cpu_load_info, &allcpus[which_logical_cpu * CPUSTATES], sizeof(long) * CPUSTATES);
+    FREE_C_HEAP_ARRAY(long, allcpus);
+#else
+    char mib[24];
```

Comment:
I might suggest "sysctl_name" or just "name" might be a more informative variable name here.  The parameter in the man page is called "sname".  Optional.

FWIW, it sounds like we could do this the same way we do FreeBSD from the NetBSD manual page.  Although I think this is simpler and am not suggesting we change at this point.

Reply:
Yes I am the master of mystical variable names :-)